File: project_euler/Problem085/Problem085.go
package project_euler

import (
    "fmt"
    "math"
)

func Problem085() {
    target := 2000000
    border := int(math.Sqrt(1 + 8 * math.Sqrt(float64(target))) - 1) >> 1
    result := 0
    pivot := target
    for i := 1; i <= border; i++ {
        length := int(math.Sqrt(1 + float64(target << 4) / float64(i) / float64(i + 1)) - 1) >> 1
        value := int(math.Min(math.Abs(float64(getCount(length, i) - target)), math.Abs(float64(getCount(length + 1, i) - target))))
        if pivot > value {
            result = length * i
            pivot = value
        }
    }
    fmt.Println(result)
}

func getCount(length int, width int) int {
    return length * (length + 1) * width * (width + 1) >> 2
}

File: project_euler/Problem085/Problem085.java
package project_euler;

public class Problem085 {
    public void solution() {
        int target = 2000000;
        int border = ((int) Math.sqrt(1 + 8 * Math.sqrt(target)) - 1) >> 1;
        int result = 0;
        int pivot = target;
        for (int i = 1; i <= border; ++i) {
            int length = ((int) Math.sqrt(1 + (double) (target << 4) / i / (i + 1)) - 1) >> 1;
            int value = Math.min(Math.abs(getCount(length, i) - target), Math.abs(getCount(length + 1, i) - target));
            if (pivot > value) {
                result = length * i;
                pivot = value;
            }
        }
        System.out.println(result);
    }

    private int getCount(int length, int width) {
        return length * (length + 1) * width * (width + 1) >> 2;
    }
}
